// @cheqd/sdk/build/cjs/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceType = exports.ISignInputs = exports.MethodSpecificIdAlgo = exports.VerificationMethods = exports.CheqdNetwork = void 0;
var CheqdNetwork;
(function (CheqdNetwork) {
    CheqdNetwork["Mainnet"] = "mainnet";
    CheqdNetwork["Testnet"] = "testnet";
})(CheqdNetwork || (exports.CheqdNetwork = CheqdNetwork = {}));
var VerificationMethods;
(function (VerificationMethods) {
    VerificationMethods["Ed255192020"] = "Ed25519VerificationKey2020";
    VerificationMethods["Ed255192018"] = "Ed25519VerificationKey2018";
    VerificationMethods["JWK"] = "JsonWebKey2020";
})(VerificationMethods || (exports.VerificationMethods = VerificationMethods = {}));
var MethodSpecificIdAlgo;
(function (MethodSpecificIdAlgo) {
    MethodSpecificIdAlgo["Base58"] = "base58btc";
    MethodSpecificIdAlgo["Uuid"] = "uuid";
})(MethodSpecificIdAlgo || (exports.MethodSpecificIdAlgo = MethodSpecificIdAlgo = {}));
exports.ISignInputs = {
    isSignInput(object) {
        return object.some((x) => 'privateKeyHex' in x);
    },
};
var ServiceType;
(function (ServiceType) {
    ServiceType["LinkedDomains"] = "LinkedDomains";
})(ServiceType || (exports.ServiceType = ServiceType = {}));
