// @cheqd/sdk/build/cjs/modules/_.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractCheqdSDKModule = void 0;
exports.instantiateCheqdSDKModule = instantiateCheqdSDKModule;
exports.instantiateCheqdSDKModuleRegistryTypes = instantiateCheqdSDKModuleRegistryTypes;
exports.instantiateCheqdSDKModuleQuerierExtensionSetup = instantiateCheqdSDKModuleQuerierExtensionSetup;
exports.applyMixins = applyMixins;
class AbstractCheqdSDKModule {
    _signer;
    methods = {};
    querier;
    _protectedMethods = ['constructor', 'getRegistryTypes', 'getQuerierExtensionSetup'];
    static registryTypes = [];
    static querierExtensionSetup = (base) => ({});
    constructor(signer, querier) {
        if (!signer) {
            throw new Error('signer is required');
        }
        if (!querier) {
            throw new Error('querier is required');
        }
        this._signer = signer;
        this.querier = querier;
    }
}
exports.AbstractCheqdSDKModule = AbstractCheqdSDKModule;
function instantiateCheqdSDKModule(module, ...args) {
    return new module(...args);
}
function instantiateCheqdSDKModuleRegistryTypes(module) {
    return module.registryTypes ?? [];
}
function instantiateCheqdSDKModuleQuerierExtensionSetup(module) {
    return module.querierExtensionSetup ?? {};
}
function applyMixins(derivedCtor, constructors) {
    let methods = {};
    constructors.forEach((baseCtor) => {
        Object.getOwnPropertyNames(baseCtor.prototype).forEach((name) => {
            const property = baseCtor.prototype[name];
            if (typeof property !== 'function' ||
                derivedCtor.hasOwnProperty(name) ||
                derivedCtor?.protectedMethods.includes(name) ||
                baseCtor.prototype?._protectedMethods?.includes(name))
                return;
            methods = { ...methods, [name]: property };
        });
    });
    return methods;
}
